lazysnap/src/main/java/com/google/accompanist/lazysnap/SnappingFlingBehavior.kt

```diff
@@ -0,0 +1,689 @@
+ *
+ * @param lazyListState The [LazyListState] to update.
+ * @param snapOffsetForItem Block which returns which offset the given item should 'snap' to.
+ * See [SnapOffsets] for provided values.
+ * @param maximumFlingDistance Block which returns the maximum fling distance in pixels.
+ * The returned value should be >= 0.
+ * @param decayAnimationSpec The decay animation spec to use for decayed flings.
+ * @param snapAnimationSpec The animation spec to use when snapping.
+ * @param endContentPadding The amount of content padding on the end edge of the lazy list
+ * in pixels (end/bottom depending on the scrolling direction).
+ */
+@ExperimentalLazySnapApi
+@Composable
+fun rememberSnappingFlingBehavior(
+    lazyListState: LazyListState,
+    snapOffsetForItem: (layoutInfo: LazyListLayoutInfo, itemInfo: LazyListItemInfo) -> Int = SnapOffsets.Center,
+    maximumFlingDistance: (LazyListLayoutInfo) -> Int = SnappingFlingBehaviorDefaults.maximumFlingDistance,
+    decayAnimationSpec: DecayAnimationSpec<Float> = rememberSplineBasedDecay(),
+    snapAnimationSpec: AnimationSpec<Float> = SnappingFlingBehaviorDefaults.snapAnimationSpec,
+    @Px endContentPadding: Int = 0,
+): SnappingFlingBehavior = remember(
+    lazyListState,
+    snapOffsetForItem,
+    maximumFlingDistance,
+    decayAnimationSpec,
+    snapAnimationSpec,
+) {
+    SnappingFlingBehavior(
+        lazyListState = lazyListState,
+        snapOffsetForItem = snapOffsetForItem,
+        maximumFlingDistance = maximumFlingDistance,
+        decayAnimationSpec = decayAnimationSpec,
+        snapAnimationSpec = snapAnimationSpec,
+    )
+}.apply {
+    this.endContentPadding = endContentPadding
+}
+
+/**
+ * Contains a number of values which can be used for the `snapOffsetForItem` parameter on
+ * [rememberSnappingFlingBehavior] and [SnappingFlingBehavior].
+ */
+@Suppress("unused")
+object SnapOffsets {
+    /**
+     * Snap offset which results in the start edge of the item, snapping to the start scrolling
+     * edge of the lazy list.
+     */
+    val Start: (LazyListLayoutInfo, LazyListItemInfo) -> Int = { _, _ -> 0 }
+
+    /**
+     * Snap offset which results in the item snapping in the center of the scrolling viewport
+     * of the lazy list.
+     */
+    val Center: (LazyListLayoutInfo, LazyListItemInfo) -> Int = { layoutInfo, itemInfo ->
+        (layoutInfo.layoutSize - itemInfo.size) / 2
+    }
+
+    /**
+     * Snap offset which results in the end edge of the item, snapping to the end scrolling
+     * edge of the lazy list.
+     */
+    val End: (LazyListLayoutInfo, LazyListItemInfo) -> Int = { layoutInfo, itemInfo ->
+        layoutInfo.layoutSize - itemInfo.size
+    }
+
+    /**
+     * Ideally this would exist on [LazyListLayoutInfo] but it doesn't right now.
+     * Raised https://issuetracker.google.com/issues/200920410 to track.
+     */
+    private val LazyListLayoutInfo.layoutSize: Int
+        get() = if (this is LazyListLayoutInfoWithContentPadding) {
```

Comment:
that is a clever idea, I like it

however I though about it more and I guess we shouldn't do it. the reason is that once we add a new property in LazyListLayoutInfo interface it will be not really a binary compatible change as any external implementations of LazyListLayoutInfo will break - they do not implement a new property. we even internally discussing using sealed interfaces for such use cases. which means we should probably not implement LazyListLayoutInfo. which means that we need to provide all the needed info as a param for the lambda blocks you have.
first approach is to just add endContentPadding as a third param, which makes it less clean, agree
second approach could be to try to abstract it away from Lazy apis and introduce our own class which contains all the needed info for calculating snap offsets. like in your current implementations you only used layout size and item size, so simple class containing those two vals will work. if you think about it in fact FlingBehavior is not a concept of lazy lists. this could be also provided for just Modifier.horizontalScroll(), and theoretically why can't we use SnappingFlingBehavior there? maybe the long term approach would be to not use the info from LazyListState directly, but map it into some new representation which SnappingFlingBehavior uses internally. this will also allow the api to be used with lazy grids, for example, as they would have their own state and layout info classes

Reply:
Yeah I agree on wanting to generalize this, but I don't think we should do it too early. The abstractions we write now may not work for the future grid APIs, etc.

I think we should stick to the `endContentPadding` thing for now (and focus on `LazyListState`), and we can work out how to generalize later. WDYT?

Reply:
yeah, whatever you think is reasonable which is not requiring implementing the interface is ok to me